torchtext/datasets/cnndm.py

```diff
@@ -0,0 +1,115 @@
+import os
+from functools import partial
+from typing import Union, Tuple
+
+from torchtext._internal.module_utils import is_module_available
+from torchtext.data.datasets_utils import (
+    _wrap_split_argument,
+    _create_dataset_directory,
+)
+
+if is_module_available("torchdata"):
+    from torchdata.datapipes.iter import (
+        FileOpener,
+        IterableWrapper,
+        OnlineReader,
+        GDriveReader,
+    )
+
+DATASET_NAME = "CNNDM"
+
+URL_LIST = {
+    "train": "https://raw.githubusercontent.com/abisee/cnn-dailymail/master/url_lists/all_train.txt",
+    "val": "https://raw.githubusercontent.com/abisee/cnn-dailymail/master/url_lists/all_val.txt",
+    "test": "https://raw.githubusercontent.com/abisee/cnn-dailymail/master/url_lists/all_test.txt",
+}
+
+STORIES_LIST = {
+    "cnn": "https://drive.google.com/uc?export=download&id=0BwmD_VLjROrfTHk4NFg2SndKcjQ",
+    "dailymail": "https://drive.google.com/uc?export=download&id=0BwmD_VLjROrfM1BxdkxVaTY2bWs",
+}
+
+PATH_LIST = {
+    "cnn": "cnn_stories.tgz",
+    "dailymail": "dailymail_stories.tgz",
+}
+
+STORIES_MD5 = {"cnn": "85ac23a1926a831e8f46a6b8eaf57263", "dailymail": "f9c5f565e8abe86c38bfa4ae8f96fd72"}
+
+_EXTRACTED_FOLDERS = {
+    "cnn": os.path.join("cnn", "stories"),
+    "daily_mail": os.path.join("dailymail", "stories"),
+}
+
+
+def _filepath_fn(root: str, source: str, _=None):
+    return os.path.join(root, PATH_LIST[source])
+
+
+# this function will be used to cache the contents of the tar file
+def _extracted_filepath_fn(root: str, source: str, t):
+    return os.path.join(root, _EXTRACTED_FOLDERS[source])
+
+
+def _filter_fn(story_fnames, x):
+    return os.path.basename(x[0]) in story_fnames
+
+
+def _get_split_list(split: str):
+    url_dp = IterableWrapper([URL_LIST[split]])
+    online_dp = OnlineReader(url_dp)
+    return online_dp.readlines().parse_cnndm_split()
```

Comment:
`parse_cnndm_split` is essentially mapping URLs to corresponding hash+'.story' right? If so, instead of creating DP, could we just create a private function and then use `map`.

Reply:
Do we have any guidance on when to create a DP to do some processing vs using map + private function? Are we only creating a DP for parsing through the dataset?

Reply:
I spoke with @parmeet offline and I think the sentiment was that if the processing is simple and is mapping some input text into some output text as is done by `parse_cnndm_split` then we should use the `map` approach. If on the otherhand the processing complex as is the case for `parse_cnndm_data` then we should convert the helper function into a datapipe. Sorry for the extra work to change this back to a private fn!